flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java

```diff
@@ -1211,7 +1231,10 @@ private boolean processFail(Throwable t, boolean isCallback, Map<String, Accumul
 				return false;
 			}
 
-			if (transitionState(current, FAILED, t)) {
+			if (!isCallback && !isLegacyScheduling()) {
```

Comment:
The new `DefaultScheduler` needs to update internal state for every task failure. This should happen via `DefaultScheduler#updateTaskExecutionState(...)`. Normally the TaskManager will invoke this method via a JobManager RPC. However, sometimes the task can be failed for other reasons that did not happen on the TaskManager. In such cases the `DefaultScheduler` should still be informed about the Task failure by invoking `vertex.getExecutionGraph().notifyTaskFailed(attemptId, t)`. We do not want to inform the `DefaultScheduler` twice and the condition attempts to model this. I am having some doubts about its correctness now. Maybe we there are other means to filter double invocations. For example, using the attemptId.

Reply:
I think it's best is to avoid the DefaultScheduler and ExecutionGraph notifying state changes to each other from the beginning.
We can let DefaultScheduler take full control of execution state changes. And thus no need for the ExecutionGraph to notify the scheduler about it. This requires a deep refactoring of current scheduling logics though, as mentioned in [this comment](https://github.com/apache/flink/pull/9663#pullrequestreview-288377060).

Reply:
I will think about whether it makes sense to extract more logic from `ExecutionGraph`. However, I think that I have found a fix for the problem: 37934d2c2930520de5587ac9c5d87eaa3e90a309